hummingbot/market/kucoin/kucoin_api_order_book_data_source.py

```diff
@@ -0,0 +1,282 @@
+        super().__init__()
+        self._symbols: Optional[List[str]] = symbols
+
+    @classmethod
+    @async_ttl_cache(ttl=60 * 30, maxsize=1)
+    async def get_active_exchange_markets(cls) -> pd.DataFrame:
+        """
+        Returned data frame should have symbol as index and include usd volume, baseAsset and quoteAsset
+        """
+        async with aiohttp.ClientSession() as client:
+
+            market_response, exchange_response = await safe_gather(
+                client.get(KUCOIN_TICKER_URL),
+                client.get(KUCOIN_SYMBOLS_URL)
+            )
+            market_response: aiohttp.ClientResponse = market_response
+            exchange_response: aiohttp.ClientResponse = exchange_response
+
+            if market_response.status != 200:
+                raise IOError(f"Error fetching Kucoin markets information. "
+                              f"HTTP status is {market_response.status}.")
+            if exchange_response.status != 200:
+                raise IOError(f"Error fetching Kucoin exchange information. "
+                              f"HTTP status is {exchange_response.status}.")
+
+            market_data = await market_response.json()
+            exchange_data = await exchange_response.json()
+
+            attr_name_map = {"baseCurrency": "baseAsset", "quoteCurrency": "quoteAsset"}
+
+            trading_pairs: Dict[str, Any] = {
+                item["symbol"]: {attr_name_map[k]: item[k] for k in ["baseCurrency", "quoteCurrency"]}
+                for item in exchange_data["data"]
+                if item["enableTrading"] == true
+            }
+
+            market_data: List[Dict[str, Any]] = [
+                {**item, **trading_pairs[item["symbol"]]}
+                for item in market_data["data"]["ticker"]
+                if item["symbol"] in trading_pairs
+            ]
+
+            # Build the data frame.
+            all_markets: pd.DataFrame = pd.DataFrame.from_records(data=market_data, index="symbol")
+            all_markets.loc[:, "USDVolume"] = all_markets.volvalue
+            all_markets.loc[:, "volume"] = all_markets.vol
+
+            return all_markets.sort_values("USDVolume", ascending=False)
+
+    async def get_trading_pairs(self) -> List[str]:
+        if not self._symbols:
+            try:
+                active_markets: pd.DataFrame = await self.get_active_exchange_markets()
+                self._symbols = active_markets.index.tolist()
+            except Exception:
+                self._symbols = []
+                self.logger().network(
+                    f"Error getting active exchange information.",
+                    exc_info=True,
+                    app_warning_msg=f"Error getting active exchange information. Check network connection."
+                )
+        return self._symbols
+
+    @staticmethod
+    async def get_snapshot(client: aiohttp.ClientSession, trading_pair: str) -> Dict[str, Any]:
+        params: str = "?symbol=" + trading_pair
```

Comment:
Use `urlencode` instead to ensure proper encoding is done on url parameters in REST API calls